File: engine.py
import json
import yfinance as yf
import pandas as pd

def search(name):
    """
        Searches the stock name through etf.txt 
        and if found return the line in txt file as an dict
        else return none

    Args:
        name (str): Stock name
    Returns:
        dict: dictionary of stock name, its leverage, and underlying assets
    """
    ticker_to_find = name.strip().upper()

    with open("etf.txt", 'r') as file:
        for line in file:
            parts = line.split(';')
            ticker = parts[0].strip()

            if ticker == ticker_to_find:
                leverage = float(parts[1].strip())

                # Parse the asset as JSON if possible, otherwise keep it as a string
                asset_str = parts[2].strip()
                try:
                    asset = json.loads(asset_str)
                except json.JSONDecodeError:
                    asset = asset_str

                # Returning the found data
                return {"ticker": ticker, "leverage": leverage, "asset": asset}
        
        # If ticker not found in the file, return error
        return None
        
def divergence(name,start_date,end_date,data):
    """
        Calculate the divergence of the leverage ETF/ETN and its
        underlying asset

    Args:
        name (str): stock name
        start_date (str): start date to grab data
        end_date (str): end date to stop grabbing data
        data (_type_): data about the underlying asset and the leverage

    Returns:
        result (float): Divergence between leverage ETF/ETN and its underlying asset
    """
    asset = 0
    etf_weight = 0
    # for leverage etf no need to calculate the weight of it
    etf = underlying(name,start_date,end_date,1)
    
   
    if isinstance(data["asset"],dict):
    
        for stock, weight in data["asset"].items():
            # for etf/etn that have underlying asset that make up the
            # percent of the etf (i.e. FNGU)
            etf_weight = underlying(stock,start_date,end_date,weight/100)
            if etf_weight == -401:
                return -401
            else:
                asset = asset + etf_weight
    else:
        asset_name = data["asset"]
        etf_weight = underlying(asset_name,start_date,end_date,1)
        
        if etf_weight == -401:
            return -401
        else:
            asset = etf_weight
        
    leverage = data["leverage"]
    result = abs(etf) - abs((leverage* asset))
    return result
    
def underlying(name,start_date,end_date,weight):
    """ Calculates the weight of the stock

    Args:
        name (str): stock name
        start_date (str): start date to grab data
        end_date (str): end date to stop grabbing data
        weight (float): weight of the asset

    Returns:
        weighted_return (float): calculated weight
    """
    stock = yf.Ticker(name)
    df = stock.history(start=start_date, end=end_date) 
    df.drop(['Dividends', 'Stock Splits','Open','High','Low','Volume'], axis=1, inplace=True)
    df['Date'] = df.index

    if df.empty or 'Close' not in df:
        return -401
    # Drop the Date as index
    df.reset_index(drop=True, inplace=True)
    start_price = df['Close'][0]
    length = len(df['Close'])
    end_price = df['Close'][length-1]
    return_percentage = (end_price - start_price) / start_price * 100
    weighted_return = return_percentage * weight
    return weighted_return

def driver(stock,start_date,end_date):
    try:
        data = search(stock)
        if data:
            return divergence(stock, start_date, end_date, data)

    except KeyError:
        return -400



#stock="FNGU"
# use try and except to see if the etf name is in the etf_data
# if not .get() should return an error
#decay_data = driver(stock, '2022-01-01', '2022-12-31')
#print(decay_data)

File: app/src/engine.js
const fs = require('fs');
const yahooFinance = require('yahoo-finance2').default;

async function search(name) {
    const tickerToFind = name.trim().toUpperCase();

    try {
        const data = fs.readFileSync('etf.txt', 'utf8');
        const lines = data.split('\n');

        for (const line of lines) {
            const parts = line.split(';');
            const ticker = parts[0].trim();

            if (ticker === tickerToFind) {
                const leverage = parseFloat(parts[1].trim());
                let asset;

                try {
                    asset = JSON.parse(parts[2].trim());
                } catch (error) {
                    asset = parts[2].trim();
                }

                return { ticker, leverage, asset };
            }
        }

        return null;
    } catch (err) {
        console.error(err);
        return null;
    }
}

async function getStockData(stock, startDate, endDate) {
    try {
        const historicalData = await yahooFinance.historical(stock, {
            period1: new Date(startDate),
            period2: new Date(endDate)
        });

        const closePrices = historicalData.map(data => data.close);
        return closePrices;
    } catch (error) {
        console.error(error);
        return null;
    }
}

async function underlying(name, startDate, endDate, weight) {
    const closePrices = await getStockData(name, startDate, endDate);
    if (!closePrices || closePrices.length === 0) {
        return -401;
    }

    const startPrice = closePrices[0];
    const endPrice = closePrices[closePrices.length - 1];
    const returnPercentage = (endPrice - startPrice) / startPrice * 100;
    return returnPercentage * weight;
}

async function divergence(name, startDate, endDate, data) {
    let asset = 0;
    let etf = await underlying(name, startDate, endDate, 1);
    if (etf === -401) return -401;

    if (typeof data.asset === 'object') {
        for (const [stock, weight] of Object.entries(data.asset)) {
            const etfWeight = await underlying(stock, startDate, endDate, weight / 100);
            if (etfWeight === -401) return -401;
            asset += etfWeight;
        }
    } else {
        const assetName = data.asset;
        const etfWeight = await underlying(assetName, startDate, endDate, 1);
        if (etfWeight === -401) return -401;
        asset = etfWeight;
    }

    const leverage = data.leverage;
    return Math.abs(etf) - Math.abs(leverage * asset);
}

async function driver(stock, startDate, endDate) {
    try {
        const data = await search(stock);
        if (data) {
            return await divergence(stock, startDate, endDate, data);
        }
    } catch (error) {
        console.error(error);
        return -400;
    }
}

// Example usage
 (async () => {
     const decayData = await driver('FNGU', '2022-01-01', '2022-12-31');
     console.log(decayData);
 })();
